Sources/Apodini/SyntaxTreeVisitor/ContextKey.swift

```diff
@@ -5,11 +5,46 @@
 //  Created by Paul Schmiedmayer on 6/26/20.
 //
 
-protocol ContextKey {
+/// A `OptionalContextKey` serves as a key definition for a `ContextNode`.
+/// Optionally it can serve a reduction logic when inserting a new value into the `ContextNode`,
+/// see `OptionalContextKey.reduce(...)`.
+/// The `OptionalContextKey` is optional in the sense that it doesn't provide a default value, meaning
+/// it may not exists on the `Context` for a given `Handler`.
+protocol OptionalContextKey {
+    /// The type of the value the `OptionalContextKey` identifies. The value MUST NOT be of type `Optional`.
     associatedtype Value
-    
-    static var defaultValue: Self.Value { get }
 
-    static func reduce(value: inout Self.Value,
-                       nextValue: () -> Self.Value)
+    /// This function can be optionally implemented to provide a reduction logic when inserting a new
+    /// value into a `ContextNode`. By default the previous value will just be overwritten.
+    /// As `OptionalContextKey` do not provide a default value, `reduce(...)` will not be called on first
+    /// insertion of a value for this `OptionalContextKey`.
+    ///
+    /// - Parameters:
+    ///   - value: The current value of the context key.
+    ///         The result of the reduction must be written into this inout parameter.
+    ///   - nextValue: The return value of the provided closure is the newly inserted value.
+    static func reduce(value: inout Self.Value, nextValue: () -> Self.Value)
+}
+
+extension OptionalContextKey {
+    static func reduce(value: inout Self.Value, nextValue: () -> Self.Value) {
+        value = nextValue()
+    }
+}
+
+
+/// A `ContextKey` is a `OptionalContextKey` with the addition of the definition of a default value.
+/// See implications of the reduction logic `OptionalContextKey.reduce(...)`.
+protocol ContextKey: OptionalContextKey {
+    /// The type of the value the `OptionalContextKey` identifies. The value MUST NOT be of type `Optional`.
+    /// The type is equal to `OptionalContextKey.Value`
+    associatedtype DefaultValue
+
+    /// The default value this `ContextKey` provides.
+    static var defaultValue: DefaultValue { get }
+}
+
+extension ContextKey {
+    // I know the compiler creates a warning about this, but using it as a type constraint isn't actually the same thing
```

Comment:
I would categorize this as "swift compiler weirdness"

If `ContextKey` was just defined like the following ...

```swift
protocol ContextKey: OptionalContextKey {
    static var defaultValue: Value { get }
}
```

... one would assume a definition like this is valid:

```swift
struct ExampleContextKey: ContextKey {
  static let defaultValue: String = "asdf"
}
```

It isn't. Although default value is defined to be type `String`, the swift compiler can't infer the type of the `associatedtype Value`  located in the parent protocol `OptionalContextKey` (I assume because the definition of the associated type isn't in the same protocol?).

My workaround then was to define the associated type `DefaultValue` and just assign it to `Value`, like dedicated in the commented code.

Like the code is right now, the above `ExampleContextKey` compiles as expected. The compiler warning wishes to refactor `ContextKey` instead to ...

```swift
protocol ContextKey: OptionalContextKey where Value == DefaultValue {
    associatedtype DefaultValue

    static var defaultValue: DefaultValue { get }
}
```

... which is seemingly not the same. First of all it is not the same as to my understanding how the `where` clause works, and indeed the `ExampleContentKey` doesn't compile anymore without explicitly adding the `typealias Value = String`.

If anyone has a better workaround for this, feel free to suggest a change 🎉

Reply:
@Supereg Thanks for the detailed explanation!

As you said an alternative to the `DefaultValue` and the extension below would be to just state 
```swift
    static var defaultValue: Self.Value { get }
```
As sated this would produce no compiler warning in this section as it is defined by the explicit declaration that `Value` is the associatedtype of `OptionalContextKey `.

As you stated the downside would be that you currently have to manually define an `typealias` like this to make the Swift compiler happy as it seems like it can't solve the type constraints across this protocol relationship at the moment:
```swift
struct OperationContextKey: ContextKey {
    typealias Value = Operation
    
    static var defaultValue: Operation = .automatic
}
```

II think this manual definition of a `typealias` would be an acceptable approach for me as this approach would reduce the number of `associatedtype`s by one and once the swift compiler gets smart enough this should be easier solvable then the current implementation with a second `associatedtype`. 
The two `associatedtype`s would point to the same type which might not be immediately visible when taking a look at the type definitions without taking a look at the externs below.

In the context of steadily reducing warnings on develop I would consider this a more likable solution as `ContextKey` and `OptionalContextKey` are both internal and not part of the public API, what do you think @Supereg @moritzschuell ?

Reply:
@Supereg @PSchmiedmayer I also don't think that having to manually define the `typealias` would be a problem. And also, it somehow feels like the less "hacky" variant to me 🤓

Reply:
Feedback incorporated with d47c205